### ISSUE TITLE
Libtorch tensor.std(0) gets confused about which overload to use (expected vector, not scalar)

### ISSUE THREAD

**Issue by BryanWBear** [pytorch/pytorch#40287]
## 🐛 Bug

Tensor.std(0) returns a scalar instead of a vector.

## To Reproduce

```
float data[] = { 1, 2, 3, 4, 5, 6 };
torch::Tensor f = torch::from_blob(data, {2, 3});
std::cout << f.mean(0) << std::endl;
// yields a 1-d tensor as expected.
// 2.5000
// 3.5000
// 4.5000
// [ CPUFloatType{3} ]
std::cout << f.std(0) << std::endl;
// yields a scalar, unexpected.
// 1.70783
// [ CPUFloatType{} ]
```

## Expected behavior

Tensor.std(0) should return a vector to be consistent with other methods. As @ptrblck pointed out in this thread https://discuss.pytorch.org/t/tensor-std-0-returns-a-scalar-instead-of-a-vector/85974, the reason this is happening is because 0 is being interpreted as the `unbiased` argument here, and suggests `f.std(true, 0)` as a workaround. However, even `f.std(true, 0)` actually collapses dimension 1, which is also unintuitive.

Ideally, these behaviors should be fixed or listed in the documentation.

## Environment

`Libtorch 1.4.0` on XCode 11.4 

## Additional context

cc @ezyang @gchanan @zou3519 @yf225 @glaringlee @bhosmer @smessmer @ljk53 @SsnL

**Comment by ptrblck** [comment]
Note that I had a typo in the suggested workaround (dim argument should be passed first).
This code snippet should yield the desired output:
```python
from torch.utils import cpp_extension


cpp_source = """
void my_fun(torch::Tensor x)
{
    std::cout << "x.std(0) " << x.std(0) << ", x.std(0, true) " << x.std(0, true) << std::endl;
    std::cout << "at::std(x, 0) " << at::std(x, 0) << ", at::std(x, 0, true) " << at::std(x, 0, true) << std::endl;
    return;
}
"""

module = torch.utils.cpp_extension.load_inline(
    name="std_test_extension",
    cpp_sources=cpp_source,
    functions="my_fun",
    verbose=True,
)

x = torch.randn(2, 3)
module.my_fun(x)

print('python ', x.std(0))
```
Output:
```python
x.std(0) 1.21999
[ CPUFloatType{} ], 
x.std(0, true)  1.2349
 0.3686
 1.7071
[ CPUFloatType{3} ]
at::std(x, 0) 1.21999
[ CPUFloatType{} ],
 at::std(x, 0, true)  1.2349
 0.3686
 1.7071
[ CPUFloatType{3} ]

python  tensor([1.2349, 0.3686, 1.7071])
```
Could you update your PyTorch version and retry the code?

**Comment by BryanWBear** [comment]
It works! Thank you.

**Comment by ezyang** [comment]
Seems like a nasty bug in Python argument parsing. Upgrading priority.

**Comment by ezyang** [comment]
The C++ overloads are:

```
CAFFE2_API Tensor std(const Tensor & self, bool unbiased=true);
CAFFE2_API Tensor std(const Tensor & self, IntArrayRef dim, bool unbiased=true, bool keepdim=false);
```

and I'm guessing 1 is directly convertible to bool, but not IntArrayRef

**Comment by peterbell10** [comment]
Fundamental (i.e. language defined) conversions have higher priority than user defined conversions in overload resolution. So while an `int -> IntArrayRef` conversion does exist, it has lower priority than `int -> bool`.

A straightforward fix should be to add an `int dim` overload which is an exact match for overload resolution.

**Comment by peterbell10** [comment]
Unfortunately, it looks like `native_functions.yaml` doesn't let you create a function taking `int`, instead it always uses `int64_t`. `std(0)` will then fail to compile because there is no preference for the `int->bool` or `int->long` conversion. Instead you need to call as `std(0L)` or `std(0LL)` on windows.

That's an improvement because you at least can't write subtly incorrect code. For `std(0)` to work either the code generation needs to be updated, or the single dimension overloads could be added into the `TensorBody.h` template manually.

Thoughts?

**Comment by ezyang** [comment]
As you say, the fully automated luxury space communism solution would be to extend the code generator to identify when there is an ambiguous overload and automatically insert an `int` overload (and I guess we should probably also add the `int64_t` overload) in the code generated C++ headers.

I think manually adding the overloads is a lot simpler. The only downside is that you can't automatically find all the places where you need to add the overload.
